# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: widget/gtk/nsWindow.cpp
# Commit: b6e830c59764
# Full Hash: b6e830c59764b9d175faf177642ab8a9d26c8139
# Author: stransky <stransky@redhat.com>
# Date: 2022-09-19 15:58:06
# Regressor Bug: 1789581
# File Overlap Count: 1
# Description:
#   Bug 1789581 [Wayland] Use new popup position at WaylandPopupFitsToplevelWindow() if popup is moved r=emilio
#   
#   If popup is moved we can't use recent popup position (mBounds) to check popup placement but
#   use requested position given at nsWindow::NativeMoveResize().
#   
# ==============================================================================

diff -r 37b28252e060 -r b6e830c59764 widget/gtk/nsWindow.cpp
--- a/widget/gtk/nsWindow.cpp	Mon Sep 19 10:39:15 2022 +0000
+++ b/widget/gtk/nsWindow.cpp	Mon Sep 19 10:39:16 2022 +0000
@@ -1859,7 +1859,7 @@
       }
       if (popup->mPopupHint == ePopupTypePanel &&
           popup->WaylandPopupIsFirst() &&
-          popup->WaylandPopupFitsToplevelWindow()) {
+          popup->WaylandPopupFitsToplevelWindow(/* aMove */ true)) {
         // Workaround for https://gitlab.gnome.org/GNOME/gtk/-/issues/1986
         // Bug 1760276 - don't use move-to-rect when popup is inside main
         // Firefox window.
@@ -2105,8 +2105,8 @@
   }
 }
 
-bool nsWindow::WaylandPopupFitsToplevelWindow() {
-  LOG("nsWindow::WaylandPopupFitsToplevelWindow()");
+bool nsWindow::WaylandPopupFitsToplevelWindow(bool aMove) {
+  LOG("nsWindow::WaylandPopupFitsToplevelWindow() move %d", aMove);
 
   GtkWindow* parent = gtk_window_get_transient_for(GTK_WINDOW(mShell));
   GtkWindow* tmp = parent;
@@ -2124,7 +2124,8 @@
   int parentHeight = gdk_window_get_height(toplevelGdkWindow);
   LOG("  parent size %d x %d", parentWidth, parentHeight);
 
-  GdkPoint topLeft = DevicePixelsToGdkPointRoundDown(mBounds.TopLeft());
+  GdkPoint topLeft = aMove ? mPopupPosition
+                           : DevicePixelsToGdkPointRoundDown(mBounds.TopLeft());
   GdkRectangle size = DevicePixelsToGdkSizeRoundUp(mLastSizeRequest);
   LOG("  popup topleft %d, %d size %d x %d", topLeft.x, topLeft.y, size.width,
       size.height);
@@ -2188,7 +2189,7 @@
     gtk_window_resize(GTK_WINDOW(mShell), size.width, size.height);
   }
 
-  if (!aMove && WaylandPopupFitsToplevelWindow()) {
+  if (!aMove && WaylandPopupFitsToplevelWindow(aMove)) {
     // Popup position has not been changed and its position/size fits
     // parent window so no need to reposition the window.
     LOG("  fits parent window size, just resize\n");